apps/axistream/src/message.hh

```diff
@@ -0,0 +1,498 @@
+#pragma once
+
+#include <type_traits>
+#include <cstdint>
+#include "hls_macros.hh"
+#include "hls_int.hh"
+#include "array.hh"
+#include "util.hh"
+
+/*
+ * Contrains a series of classes to support interpretation of bit vectors.
+ *
+ * Example usage:
+ * Consider the C struct:
+ * struct MyStruct {
+ *     uint64_t a;
+ *     int32_t b;
+ * };
+ *
+ * This struct can be represented in the message framework as:
+ * struct A : Int<64, Sign::UNSIGNED, Endianness::LITTLE> {};
+ * struct B : Int<32, Sign::SIGNED, Endianness::LITTLE> {};
+ * using MyStruct = Struct<A, B>;
+ *
```

Comment:
OK, so fundamentally, this seems very similar to the old technique. A field of a header is represented using a named type.  You're just using variadic templates directly to make things work, instead of mpl::vector<>?

Reply:
Right - this was all inspired by your original design. Using the variardic template instead of `mpl::vector` is one key difference. Another is using inheritance to define fields, instead of using `mpl::string`.

Reply:
There is also a richer message type system now. You can have nesting, for example:
```c++
struct Field1 : MyStruct {};
struct Field2 : MyStruct {};
using MyStructPair = Struct<Field1, Field2>;
```

Enums are also supported, as are unions.